dag_in_context/src/add_context.rs

```diff
@@ -143,12 +228,16 @@ impl Expr {
             Expr::Arg(ty, _oldctx) => RcExpr::new(Expr::Arg(ty.clone(), context_to_add)),
             // create new contexts for let, loop, and if
             Expr::DoWhile(inputs, pred_and_body) => {
+                let placeholder = cache.new_placeholder();
```

Comment:
I think this code-block could really use a comment (I guess basically the summary I put in a comment above). I think it will be important to capture *why* we need to do all this, especially since the way we had it before seems simpler, and if we later forget all the shadow realm problems, we might accidentally reintroduce those problems.